File: golang/subdomain_visit.go
func subdomainVisits(cpdomains []string) []string {
	domainFreq := map[string]int{}

	for _, cpdomain := range cpdomains {
		split1 := strings.Split(cpdomain, " ")
		count, _ := strconv.Atoi(split1[0])
		domain := split1[1]

		// split parts and pop from front until empty
		parts := strings.Split(domain, ".")
		for len(parts) > 0 {
			// add joined domain
			domainFreq[strings.Join(parts, ".")] += count

			// remove the front part
			parts = append([]string{}, parts[1:]...)
		}
	}

	// build result
	result := []string{}
	for domain, count := range domainFreq {
		result = append(result, strconv.Itoa(count)+" "+domain)
	}

	return result
}

File: typescript/subdomain_visit.ts
function subdomainVisits(cpdomains: string[]): string[] {
	let freq = new Map<string, number>()

	cpdomains.forEach(cpdomain => {
		let cp = cpdomain.split(' ')
		let count = parseInt(cp[0])
		let domain = cp[1]

		while (domain.length != 0) {
			freq.set(domain, count + (freq.get(domain) ?? 0)) // default to count if undefined
			domain = domain.split('.').slice(1).join('.') // convert to array, remove first index, convert to string
		}
	})

	// populate result string interpolation
	let result = []
	for (let [domain, count] of freq.entries()) {
		result.push(`${count} ${domain}`)
	}
	
	return result
}